### ISSUE TITLE
SPOE next generation

### ISSUE THREAD

**Issue by capflam** [haproxy/haproxy#2502]
### Your Feature Request

The current implementation of the SPOE was deprecated for the 3.0 (https://www.mail-archive.com/haproxy@formilux.org/msg44680.html).  Idea was not to remove the feature but to rework it. The engine is now too outdated to make it evolve and a new one must be implemented. Mainly to be based on recently implemented core features (muxes, connection sharing...) but also to be adapted to real usage. When it was developed, the SPOE was designed with the payload streaming in mind. But it was never implemented because of the feature's complexity. Because of that, the current design is overkill and make the SPOE even harder to maintain.

With this feature request, we will try to describe the base of the next generation of the SPOE. 

### What are you trying to do?

First of all, as far as possible, the protocol (SPOP) should be kept compatible with existing agent implementations. At first glance, it seems feasible. Then the streaming must be kept out of the scope of the study. This way we will be able to simplify the design and probably the agent implementations. Finally the asynchronous mode (the ability to send a request on a connection to eventually receive the response on another one) will be removed. It was most probably never used and it is too complex to be kept in the new design.

For the 3.0, the current implementation will be kept as deprecated with a warning to notify users some configuration changes should be expected in future versions. In same time, a SPOP mode for backends will be introduced. This should ease the introduction of a SPOP multiplexer in the 3.1. Fragmented frames will no longer be used by HAProxy. Without the payload streaming, it is overkill and almost useless. Some new configuration parameters will probably be introduced to prepare the refactoring but hard to be specific for now.

For the 3.1, the connection sharing mechanism must be reviewed to support SPOP connections (or more generally any type of connection ?). For now, only HTTP connections are concerned. And a SPOP multiplexer must then be implemented. It is still foggy but the HTX should be evaluated to represent SPOP messages exchanged between the SPOE applet and the SPOP multiplexer.  It may be a good way to support other protocols with the same engine (gRPC for instance). The SPOE applet must be rewritten. It is of course a quick sum up of what should be done. Only the main milestones are mentioned.  But it is a good start for discussion.

### Output of `haproxy -vv`

```plain
HAProxy version 3.0-dev5-8674f6-59 2024/03/25 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.7.6-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Feb 23 18:27:29 UTC 2024 x86_64
```

**Comment by git001** [comment]
Let me add some feature request for the redesign phase.

It would be very helpful to be able to scan the body of a request via SPOE. Such a request comes from time to time and this is one of the limiting factor for the current implementation, from my point of view.

**Comment by wtarreau** [comment]
That was definitely part of the initial design and it requires many more changes and added a huge amount of complexity for something that was never completed. One of the approaches that we'll follow on the redesign is to make the engine better, more reliable, faster, observable etc and this passes by first being simpler. Simpler definitely implies getting rid of the streaming attempts.

I know it can sound a bit sad but let's be honest, it has had very very little traction over the last 7-8 years and is responsible for most of the complexity. Maybe that would be revisited in the future but for now one of the plans are to not waste precious time and effort on this low-demand feature.

**Comment by git001** [comment]
Well I don't agree with that statement.

> I know it can sound a bit sad but let's be honest, it has had very very little traction over the last 7-8 years and is responsible for most of the complexity. Maybe that would be revisited in the future but for now one of the plans are to not waste precious time and effort on this low-demand feature.

There was and is still are use cases where the body is very helpful for modules like the WAF's with mod_security. There is already a solution for that [Coraza SPOA - HAProxy Web Application Firewall](https://github.com/corazawaf/coraza-spoa) which could benefit from the feature to scan also the Body.

Maybe the streaming of the body could be limited to some size and do not stream the body but send some amount of data with a fixed size.

What I want to highlight is that at this redesign there should be not a design decision which makes the stream of the body almost impossible for the future.

I'm looked into this line and asked my self:
* Could this body handling reliable work?
* Is this already a working solution?

https://github.com/DropMorePackets/haproxy-go/blob/master/spop/example/header-to-body/engine.cfg#L13
```
    args id=unique-id src-ip=src method=method path=path query=query version=req.ver headers=req.hdrs body=req.body
```

**Comment by wtarreau** [comment]
From my understanding Coraza can already scan the body contained in the first buffer. I'm not saying this is the panacea but if they developed it as an SPOA it means they do see some value in doing this. We've also seen two other WAFs working this way and being used (mod_security and mod_defender).

What you mean by "not stream the body but send some amount of data with a fixed size" is exactly what is currently being done and what will be re-implemented. There's no reason to change this.

What is a total pain is to stream the data before header processing, hence before a routing decision is taken, and re-inject them in return and pass them to the next stage -- all that using the same buffer for in and out. Even the limited forwarding that we originally envisioned inline (e.g. image recompression) is a pain to deal with: how do you adjust a response header to match what you're doing with the contents for example. All of these initial intents significantly contributed to the complexity of the current implementation while still making it insufficient.

And I'm not even speaking about all the design discussions about applet architecture simplification etc that often ended up with "wait wait we can't do that, there's SPOE". My feeling was that it was almost unused and that we could just get rid of it, even if it annoys a few users, because in the end everyone pays the consequences of maintaining it by delayed features and more complex bugs in the core. Given that there has been concerns and objections we discussed this and came to an agreement to prioritize a re-implementation but now the primary focus must be maintainability, and the immediate second one is to address all the design choices that were causing trouble (including the painful idle connection management that didn't support inter-thread sharing, and the LB that was only applying to connection creation instead of messages).

I think we'll get the best of both worlds: same filtering/processing features as right now, lower resource usage for agents, much better load balancing, higher reliability, higher performance and no more maintenance hell (i.e. no longer dream of getting rid of it).

**Comment by git001** [comment]
> What is a total pain is to stream the data before header processing, hence before a routing decision is taken, and re-inject them in return and pass them to the next stage -- all that using the same buffer for in and out. Even the limited forwarding that we originally envisioned inline (e.g. image recompression) is a pain to deal with: how do you adjust a response header to match what you're doing with the contents for example. All of these initial intents significantly contributed to the complexity of the current implementation while still making it insufficient.

Ah okay, got you. I haven't thought that this would anybody need, but I'm in my own bubble :smile: .

To clarify it a bit about the wording.
* The "streaming Body" means here to re-inject the body after some processing, right?
* The "handling Body" means here to send the body do SPOA do with that what every wanted and set some variables for further HAProxy processing, right?

**Comment by wtarreau** [comment]
The "streaming body" would mean that you'd have to send all of it to the agent with the agent sending it back. In terms of communication it would have the form of a "T", with client->ha, ha->agent, agent->ha, ha->server.

What's currently supported is what you mentioned, req.body is sent to the agent without having to leave the internal buffers, and the agent sends its verdict, variables etc. And we intend to keep this, of course!

**Comment by agirot** [comment]
@wtarreau :wave: I am just curious, why rework SPOE and not try to implement proxy-wasm (https://github.com/proxy-wasm/spec) which now seems to be the standard way to add plugin/hook in proxy ecosystem (added in envoy, kong, traefik,ats...). Haproxy could take advantage of plugins already created for other proxies, it's seem pretty cool ?

https://github.com/haproxy/haproxy/issues/1482

**Comment by wtarreau** [comment]
It seems totally orthogonal. I would like on the long term to support what others also support, to try to unify the ecosystems. But here we observed some resistance from some users who had invested a lot of time on their SPOA and it just seems that we underestimated the amount of code already deployed. In fact it seems that for many of them it works well enough that they never communicate about it. And when you think about it it makes sense, they implement their own stuff internally using it, and only have to discuss them publicly if the engine doesn't work, or when we mark it deprecated 4 months after announcing it.

Thus we discussed and decided that OK we're not going to leave them in the dust, but since we can't really pursue on the existing architecture either, we'll dedicate some of the core team's time to reimplement a better and more durable architecture that will not only stop always being in the way when we try to fix stuff, but will also bring them the performance and reliability that's currently missing in the aging architecture.

Then after that it will still be time to study what is commonly accepted as "the norm" for other projects. For me what's being described here as proxy-wasm is how proxies should interact with their own wasm virtual machine, just as we have with Lua right now. And of course, if we'd eventually go down that route it would be super important to be compatible with that. But we may also one day have to support existing external agents that are compatible with other proxies. Some stuff might already work on HTTP or gRPC nowadays for example. As an illustration, before SPOP we had a thought about IMAP which used to be the norm 20 years ago. What's important is to make sure that if others unite around a technology, we're not the only ones not using it. It causes double work for implementers, and fragmentation for users, that's not great.

**Comment by zino7825** [comment]
Spoa is a really necessary function, so you need to think about next negation.

We are using so many different types of spoa.

Authentication module, mirroring, mod_security, etc..
I think it's a necessary function.

I tried implementing it with lua, but the performance was not good.

**Comment by wtarreau** [comment]
That's exactly why we're going to reimplement a new SPOE that will remain compatible with your existing SPOA.

**Comment by mcarbonneaux** [comment]
the streaming of the entire request body and response body as filter is very important missing feature of spoa, not only for waf, but most of the time to make correction of the traffic content.

you say that you will extend with grpc support, why not use it to do that like is done with envoy external processing ? 
https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter 

or simply implement the same protocol than envoy... to be abel tu reuse external processor already developped for envoy....

**Comment by ronyrv13** [comment]
Hello @capflam 
I'm using HAProxy SPOE with a custom Go agent (using the [negalus/haproxy-spoe-go](https://github.com/negasus/haproxy-spoe-go) package) to handle HTTP responses, including streaming responses (such as OpenAI API's streaming output). 
However, during response streaming, HAProxy sends the first and last chunk but then disconnects, showing the following error:

`error: handle worker: error send AgentDisconnect frame: write unix /tmp/spoe-agent.sock->@: write: broken pipe`

Even those chunks of data is also not complete.

Is this expected behavior when using SPOE with HTTP streaming, or could this be a bug? Are there recommended configurations or steps to handle streaming properly with HAProxy SPOE?
Any input on this will be a great help.

Thanks

**Comment by capflam** [comment]
@ronyrv13, Streaming is not supported with the SPOE. At best, only the first buffer can be sent to the agent. in HTTP, you can wait for the message payload before starting the analysis. In this case, all received data, up to a buffer size minus the reserve (around 15k by default) will be available. If the message is small enough, it is ok.

Now, I don't know how `haproxy-spoe-go` agent is working. So It is hard to answer. The project is not pretty active. On the other hand, I have no information on your setup (haproxy version, configuration ...). If you are sure your go agent should work as expected and the problem is on HAProxy side, you can open an new issue indeed.

**Comment by ErmakovDmitriy** [comment]
Hello, could you, please, include in the future version of the SPOE-NG a possibility to pass headers, cookies etc from HAProxy and pass back `actions` (redirect, deny request etc)?

It is used in an OpenID authentication agent for HAProxy:

- https://github.com/criteo/haproxy-spoe-auth/blob/master/resources/haproxy/spoe-auth.conf - HAProxy configuration, part 1
- https://github.com/criteo/haproxy-spoe-auth/blob/master/resources/haproxy/haproxy.cfg#L22 - HAProxy configuration, part 2

**Comment by wtarreau** [comment]
This feedback is useful. We've seen some configs that had many set-var() rules just to copy headers, and we also thought that maybe having a simpler way to pass a whole headers block back and forth would be convenient. For actions, that could be trickier however, as they generally take arguments and all of these are placed into immutable config blocks at parsing time. There are also risks in accepting any action from an SPOA (think that some might not be under your own authority but be hosted by a service provider). But this might deserve some thoughts at least.

**Comment by ErmakovDmitriy** [comment]
Thank you.

True, the actions might be unsafe but, probably, it is okay not to implement them, if we can do something like:

```
# Provided that the sess.auth.is_authenticated can be set by an SPOE agent
# based on some processing, maybe checking cookies, headers, an external database etc.
acl authenticated var(sess.auth.is_authenticated) -m bool

http-request accept if authenticated
use_backend my_app if authenticated

http-request redirect location https://authserver.example.com/auth if !authenticated
```

**Comment by ronyrv13** [comment]
@capflam, Thank you for the clarification. So, as you mentioned, streaming is not supported with SPOE, which means it may not be the right fit for my use case.

**Use Case**:  
I am trying to use SPOE to capture specific values from the response body. In particular, I need to extract the "usage" object from OpenAI's API response, which contains information like:

```
"usage": {"completion_tokens": 95, "prompt_tokens": 56, "total_tokens": 151}
```

When I make a **non-streaming** request, everything works as expected. SPOE makes an asynchronous call to the SPOA, and I can capture these values and set variables that I later use for logging.

However, when making a **streaming** request to the OpenAI API, the response is returned as data-only [server-sent events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events), and I receive the "usage" object in the last chunk of the response. I need to capture this value **without buffering the entire response** or breaking the streaming process.

I considered using even though i think `json_query` buffers the response :

```
http-response set-var(txn.prompt_tokens) res.body,json_query('$.usage.prompt_tokens','int')
```

But it also doesn't work well with chunked responses may be because the required value is in the final chunk.

At this point, I'm unsure how to handle this situation. My preference is to continue using HAProxy for this, as I’ve always relied on it for load balancing and proxying needs. However, I’m not sure how to achieve this functionality with streaming responses. 

Do you have any suggestions or alternative approaches that could work with HAProxy and SPOE, or perhaps a different configuration that might help capture these values without disrupting the streaming?

Thanks again for your help and insights.

**Comment by capflam** [comment]
@ronyrv13, indeed, in your case, the SPOE without the streaming is not suitable. If your workload permits it, you can evaluate the lua-filters. You will be able to filter all the request payload this way. Of course, this comes with a cost. But if it is acceptable, this should work.

**Comment by ygkat** [comment]
@capflam

I would like to propose some improvements for SPOE.

1. It would be great if we could configure timeout processing for each spoe-message.
```haproxy
spoe-group ip
    messages ip

spoe-message ip
    timeout processing 100ms

    args src=src hdrs=req.hdrs
```

2. There is no spoe queue timeout, and setting a queue timeout for the spoe-agent backend doesn't have any effect. Are there any plans for improvement?
```haproxy
backend spoa
    mode tcp
    timeout connect 2s
    timeout server 10s
    timeout queue 10ms

    option spop-check
    server spoa_server 1.1.1.1:8080 check

# spoe log qT=50ms
sid=1546555 st=1 0/50/-1/-1/80 40/40 0/39 1901/1546544
```

**Comment by capflam** [comment]
The SPOE was rewritten for the 3.1. Messages are no longer pushed in a dedicated queue to be handled by a pool of applets. SPOE groups or events are now more close to what is performed for HTTP transactions. When a SPOE group or a SPOE event is processed, we request a new connection to the agent via the normal way.

**Comment by capflam** [comment]
About a timeout per-message, it is not applicable because several messages can be grouped in the same frame. For instance several messages in the same SPOE group. However, it may be applicable to a group or an event. For a SPOE group, I guess it is not too hard to achieve, for SPOE events, it is a bit tedious but probably feasible.

**Comment by ygkat** [comment]
I kindly ask for your positive consideration.

There are many advantages to calling `spoe-group` at the desired time.

When not using spoe-group, we use `on-frontend-http-request`, but the spoe is called before the rules (e.g., ACL) set on the frontend are executed, making it less flexible to use.
Additionally, since `spoe-message` does not support timeout processing, multiple spoe engine sections are required for a single spoa backend.

Thank you.
